File: python-backend/app/services/data_metrics_config_service.py
"""Data metrics config service – async SQLAlchemy port of DataMetricsConfigService.ts."""

from __future__ import annotations

import json
import time
from typing import Any

from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from app.models.data_metrics_config import DataMetricsConfig

# ---------------------------------------------------------------------------
# Default work-status configuration (mirrors workStatus.config.ts)
# ---------------------------------------------------------------------------

DEFAULT_THRESHOLDS: dict[str, int] = {
    "relaxed": 6,
    "normal": 10,
    "busy": 15,
    "superCrazy": 20,
}
DEFAULT_OVERTIME_HOUR: int = 19

WORK_STATUS_LABELS: dict[str, str] = {
    "relaxed": "轻松",
    "normal": "正常",
    "busy": "忙碌",
    "crazy": "疯狂",
    "overtime": "加班",
    "superCrazyOvertime": "超级疯狂加班",
}

WORK_STATUS_COLORS: dict[str, str] = {
    "relaxed": "green",
    "normal": "blue",
    "busy": "orange",
    "crazy": "red",
    "overtime": "red",
    "superCrazyOvertime": "magenta",
}

VALID_ANT_COLORS: set[str] = {
    "default",
    "processing",
    "success",
    "error",
    "warning",
    "magenta",
    "red",
    "volcano",
    "orange",
    "gold",
    "lime",
    "green",
    "cyan",
    "blue",
    "geekblue",
    "purple",
}


class DataMetricsConfigService:
    """Manage work-status thresholds / overtime-hour / labels / colors."""

    # ------------------------------------------------------------------
    # helpers
    # ------------------------------------------------------------------

    @staticmethod
    def get_default_config() -> dict[str, Any]:
        now = int(time.time() * 1000)
        return {
            "id": 0,
            "thresholds": DEFAULT_THRESHOLDS.copy(),
            "overtimeHour": DEFAULT_OVERTIME_HOUR,
            "labels": WORK_STATUS_LABELS.copy(),
            "colors": WORK_STATUS_COLORS.copy(),
            "createdAt": now,
            "updatedAt": now,
        }

    # ------------------------------------------------------------------
    # read
    # ------------------------------------------------------------------

    @staticmethod
    async def get_config(db: AsyncSession) -> dict[str, Any]:
        stmt = (
            select(DataMetricsConfig)
            .order_by(DataMetricsConfig.created_at.desc())
            .limit(1)
        )
        result = await db.execute(stmt)
        row = result.scalar_one_or_none()
        if row is None:
            return DataMetricsConfigService.get_default_config()

        return {
            "id": row.id,
            "thresholds": json.loads(row.thresholds),
            "overtimeHour": row.overtime_hour,
            "labels": json.loads(row.labels),
            "colors": json.loads(row.colors),
            "createdAt": int(row.created_at),
            "updatedAt": int(row.updated_at),
        }

    # ------------------------------------------------------------------
    # write
    # ------------------------------------------------------------------

    @staticmethod
    async def update_config(
        db: AsyncSession,
        thresholds: dict[str, int],
        overtime_hour: int,
        labels: dict[str, str],
        colors: dict[str, str],
    ) -> dict[str, Any]:
        # validate thresholds increasing
        if (
            thresholds["relaxed"] >= thresholds["normal"]
            or thresholds["normal"] >= thresholds["busy"]
            or thresholds["busy"] >= thresholds["superCrazy"]
        ):
            raise ValueError("阈值必须递增：relaxed < normal < busy < superCrazy")

        # validate overtime hour
        if overtime_hour < 0 or overtime_hour > 23:
            raise ValueError("加班时间阈值必须在 0-23 之间")

        # validate colors
        for status, color in colors.items():
            if color not in VALID_ANT_COLORS:
                raise ValueError(f"无效的颜色值: {color}，状态: {status}")

        now = int(time.time() * 1000)

        stmt = (
            select(DataMetricsConfig)
            .order_by(DataMetricsConfig.created_at.desc())
            .limit(1)
        )
        result = await db.execute(stmt)
        existing = result.scalar_one_or_none()

        if existing is not None:
            existing.thresholds = json.dumps(thresholds, ensure_ascii=False)
            existing.overtime_hour = overtime_hour
            existing.labels = json.dumps(labels, ensure_ascii=False)
            existing.colors = json.dumps(colors, ensure_ascii=False)
            existing.updated_at = now
            await db.flush()
            return {
                "id": existing.id,
                "thresholds": thresholds,
                "overtimeHour": overtime_hour,
                "labels": labels,
                "colors": colors,
                "createdAt": int(existing.created_at),
                "updatedAt": int(existing.updated_at),
            }

        new_row = DataMetricsConfig(
            thresholds=json.dumps(thresholds, ensure_ascii=False),
            overtime_hour=overtime_hour,
            labels=json.dumps(labels, ensure_ascii=False),
            colors=json.dumps(colors, ensure_ascii=False),
            created_at=now,
            updated_at=now,
        )
        db.add(new_row)
        await db.flush()
        return {
            "id": new_row.id,
            "thresholds": thresholds,
            "overtimeHour": overtime_hour,
            "labels": labels,
            "colors": colors,
            "createdAt": int(new_row.created_at),
            "updatedAt": int(new_row.updated_at),
        }

    # ------------------------------------------------------------------
    # init
    # ------------------------------------------------------------------

    @staticmethod
    async def init_default_config(db: AsyncSession) -> None:
        stmt = select(DataMetricsConfig).limit(1)
        result = await db.execute(stmt)
        if result.scalar_one_or_none() is not None:
            return
        default = DataMetricsConfigService.get_default_config()
        await DataMetricsConfigService.update_config(
            db,
            thresholds=default["thresholds"],
            overtime_hour=default["overtimeHour"],
            labels=default["labels"],
            colors=default["colors"],
        )

File: backend/src/services/DataMetricsConfigService.ts
import { prisma } from '../db/client.js';
import { defaultWorkStatusConfig, workStatusLabels, workStatusColors, type WorkStatusConfig, type WorkStatus } from '../config/workStatus.config.js';
import { cacheService } from './CacheService.js';

export interface DataMetricsConfig {
  id: number;
  thresholds: WorkStatusConfig['thresholds'];
  overtimeHour: number;
  labels: Record<WorkStatus, string>;
  colors: Record<WorkStatus, string>;
  createdAt: number;
  updatedAt: number;
}

export class DataMetricsConfigService {
  /**
   * 获取默认配置
   */
  getDefaultConfig(): DataMetricsConfig {
    return {
      id: 0,
      thresholds: defaultWorkStatusConfig.thresholds,
      overtimeHour: defaultWorkStatusConfig.overtimeHour,
      labels: workStatusLabels,
      colors: workStatusColors,
      createdAt: Date.now(),
      updatedAt: Date.now()
    };
  }

  /**
   * 获取当前数据指标配置（如果没有则返回默认值）
   */
  async getConfig(): Promise<DataMetricsConfig> {
    const config = await prisma.dataMetricsConfig.findFirst({
      orderBy: { createdAt: 'desc' }
    });

    if (!config) {
      return this.getDefaultConfig();
    }

    return {
      id: config.id,
      thresholds: JSON.parse(config.thresholds),
      overtimeHour: config.overtimeHour,
      labels: JSON.parse(config.labels),
      colors: JSON.parse(config.colors),
      createdAt: Number(config.createdAt),
      updatedAt: Number(config.updatedAt)
    };
  }

  /**
   * 更新数据指标配置
   */
  async updateConfig(config: {
    thresholds: WorkStatusConfig['thresholds'];
    overtimeHour: number;
    labels: Record<WorkStatus, string>;
    colors: Record<WorkStatus, string>;
  }): Promise<DataMetricsConfig> {
    const now = BigInt(Date.now());

    // 验证阈值是否递增
    const { thresholds } = config;
    if (
      thresholds.relaxed >= thresholds.normal ||
      thresholds.normal >= thresholds.busy ||
      thresholds.busy >= thresholds.superCrazy
    ) {
      throw new Error('阈值必须递增：relaxed < normal < busy < superCrazy');
    }

    // 验证加班时间阈值
    if (config.overtimeHour < 0 || config.overtimeHour > 23) {
      throw new Error('加班时间阈值必须在 0-23 之间');
    }

    // 验证颜色值（Ant Design Tag 支持的颜色）
    const validColors = ['default', 'processing', 'success', 'error', 'warning', 'magenta', 'red', 'volcano', 'orange', 'gold', 'lime', 'green', 'cyan', 'blue', 'geekblue', 'purple'];
    for (const [status, color] of Object.entries(config.colors)) {
      if (!validColors.includes(color)) {
        throw new Error(`无效的颜色值: ${color}，状态: ${status}`);
      }
    }

    // 检查是否存在配置
    const existing = await prisma.dataMetricsConfig.findFirst({
      orderBy: { createdAt: 'desc' }
    });

    if (existing) {
      // 更新现有配置
      const updated = await prisma.dataMetricsConfig.update({
        where: { id: existing.id },
        data: {
          thresholds: JSON.stringify(config.thresholds),
          overtimeHour: config.overtimeHour,
          labels: JSON.stringify(config.labels),
          colors: JSON.stringify(config.colors),
          updatedAt: now
        }
      });

      await cacheService.invalidate('metrics');
      await cacheService.invalidate('commits');
      return {
        id: updated.id,
        thresholds: config.thresholds,
        overtimeHour: config.overtimeHour,
        labels: config.labels,
        colors: config.colors,
        createdAt: Number(updated.createdAt),
        updatedAt: Number(updated.updatedAt)
      };
    } else {
      // 创建新配置
      const created = await prisma.dataMetricsConfig.create({
        data: {
          thresholds: JSON.stringify(config.thresholds),
          overtimeHour: config.overtimeHour,
          labels: JSON.stringify(config.labels),
          colors: JSON.stringify(config.colors),
          createdAt: now,
          updatedAt: now
        }
      });

      await cacheService.invalidate('metrics');
      await cacheService.invalidate('commits');
      return {
        id: created.id,
        thresholds: config.thresholds,
        overtimeHour: config.overtimeHour,
        labels: config.labels,
        colors: config.colors,
        createdAt: Number(created.createdAt),
        updatedAt: Number(created.updatedAt)
      };
    }
  }

  /**
   * 初始化默认配置（如果不存在）
   */
  async initDefaultConfig(): Promise<void> {
    const existing = await prisma.dataMetricsConfig.findFirst();
    if (existing) {
      return; // 已存在配置，不需要初始化
    }

    const defaultConfig = this.getDefaultConfig();
    await this.updateConfig({
      thresholds: defaultConfig.thresholds,
      overtimeHour: defaultConfig.overtimeHour,
      labels: defaultConfig.labels,
      colors: defaultConfig.colors
    });
  }
}
